numpy/_core/src/multiarray/multiarraymodule.h

```diff
@@ -1,26 +1,228 @@
+     * References to items obtained via an import at module initialization
+     */
+    PyObject *AxisError;
+    PyObject *ComplexWarning;
+    PyObject *DTypePromotionError;
+    PyObject *TooHardError;
+    PyObject *VisibleDeprecationWarning;
+    PyObject *_CopyMode;
+    PyObject *_NoValue;
+    PyObject *_ArrayMemoryError;
+    PyObject *_UFuncBinaryResolutionError;
+    PyObject *_UFuncInputCastingError;
+    PyObject *_UFuncNoLoopError;
+    PyObject *_UFuncOutputCastingError;
+    PyObject *math_floor_func;
+    PyObject *math_ceil_func;
+    PyObject *math_trunc_func;
+    PyObject *math_gcd_func;
+    PyObject *os_PathLike;
+    PyObject *os_fspath;
+
+    /*
+     * stores sys.flags.optimize as a long, which is used in the add_docstring
+     * implementation
+     */
+    long optimize;
+
+    /*
+     * LUT used by unpack_bits
+     */
+    union {
+        npy_uint8  bytes[8];
+        npy_uint64 uint64;
+    } unpack_lookup_big[256];
+
+    /*
+     * Used in the __array__ internals to avoid building a tuple inline
+     */
+    PyObject *kwnames_is_copy;
+
+    /*
+     * Used in __imatmul__ to avoid building tuples inline
+     */
+    PyObject *axes_1d_obj_kwargs;
+    PyObject *axes_2d_obj_kwargs;
+
+    /*
+     * Used for CPU feature detection and dispatch
+     */
+    PyObject *cpu_dispatch_registry;
+
+    /*
+     * A look-up table to recover integer type numbers from type characters.
+     *
+     * See the _MAX_LETTER and LETTER_TO_NUM macros in arraytypes.c.src.
+     *
+     * The smallest type number is ?, the largest is bounded by 'z'.
+     */
+    npy_int16 _letter_to_num['z' + 1 - '?'];
+
+    /*
+     * references to ArrayMethod implementations that are cached
+     * to avoid repeatedly creating them
+     */
+    PyObject *VoidToGenericMethod;
+    PyObject *GenericToVoidMethod;
+    PyObject *ObjectToGenericMethod;
+    PyObject *GenericToObjectMethod;
+} npy_ma_static_data_struct;
+
+
+/*
+ * A struct storing thread-unsafe global state for the _multiarray_umath
+ * module. We should refactor so the global state is thread-safe,
+ * e.g. by adding locking.
+ */
+typedef struct npy_ma_thread_unsafe_state_struct {
+    /*
+     * Cached references to objects obtained via an import. All of these are
+     * can be initialized at any time by npy_cache_import.
+     *
+     * Currently these are not initialized in a thread-safe manner but the
+     * failure mode is a reference leak for references to imported immortal
+     * modules so it will never lead to a crash unless users are doing something
+     * janky that we don't support like reloading.
+     *
+     * TODO: maybe make each entry a struct that looks like:
+     *
+     *      struct {
+     *          atomic_int initialized;
+     *          PyObject *value;
+     *      }
+     *
+     * so the initialization is thread-safe and the only possibile lock
+     * contention happens before the cache is initialized, not on every single
+     * read.
```

Comment:
Yeah, see https://github.com/numpy/numpy/pull/26430 for that. After this gets merged my next step is to integrate that into the argparse cache and this cache to make them thread-safe.

Reply:
So, I think the question remains a bit how many of these we actually want to centralize.  For strings and imports, I think it makes sense for sure.  But for the `ArrayMethods` used in the cast for example I would prefer to keep the code where it is.
And at that point, I am not sure it would be better to call the function from the main init function or rather just lock there.

Reply:
> And we will also do this for the argparse things, or do you aim to never do any init locally to avoid having to remember to explicitly lock there?

But yes, the argparse cache macro will also be updated like in #26430, the only major difference is the global lock I added there to handle locking during cache initialization would live in the global state struct. But maybe not if you're sour on that idea and it can just live in a spot local to the module like in #26430 as-is.

Reply:
I just meant that unless we have no localized init, there is maybe little reason to avoid it.